kong/clustering/control_plane.lua

```diff
@@ -196,128 +253,271 @@ do
-    np = node_plugins[j]
+  for _, plugin in ipairs(self.plugins_list) do
+    local name = plugin.name
+    local cp_plugin = self.plugins_map[name]
+    local dp_plugin = dp_plugin_map[name]
 
-    if p.name ~= np.name then
-      goto continue
+    if not dp_plugin then
+      if cp_plugin.version then
+        ngx_log(ngx_WARN, name, " plugin ", cp_plugin.version, " is missing from data plane", dp)
+      else
+        ngx_log(ngx_WARN, name, " plugin is missing from data plane", dp)
+      end
+
+    else
+      if cp_plugin.version and dp_plugin.version then
+        local msg = "data plane " .. name .. " plugin version " .. dp_plugin.version ..
+                    " is different to control plane plugin version " .. cp_plugin.version
+
+        if cp_plugin.major ~= dp_plugin.major then
+          ngx_log(ngx_WARN, msg, cp_plugin.version, dp)
+
+        else
+          local difference = math.abs(cp_plugin.minor - dp_plugin.minor)
+          if difference == 1 then
+            ngx_log(ngx_DEBUG, msg, dp)
+          elseif difference == 2 then
+            ngx_log(ngx_INFO, msg, dp)
+          else
+            ngx_log(ngx_NOTICE, msg, dp)
+          end
+        end
+
+      elseif dp_plugin.version then
+        ngx_log(ngx_NOTICE, "data plane ", name, " plugin version ", dp_plugin.version,
+                            " has unspecified version on control plane", dp)
+
+      elseif cp_plugin.version then
+        ngx_log(ngx_NOTICE, "data plane ", name, " plugin version is unspecified, ",
+                            "and is different to control plane plugin version ",
+                            cp_plugin.version, dp)
+      end
     end
+  end
+
+  return true, nil, CLUSTERING_SYNC_STATUS.NORMAL
+end
 
-    -- ignore plugins without a version (route-by-header is deprecated)
-    if p.version and np.version then
-      -- major/minor check that ignores anything after the second digit
-      local major_minor_p = p.version:match("^(%d+%.%d+)") or "not_a_version"
-      local major_minor_np = np.version:match("^(%d+%.%d+)") or "still_not_a_version"
-
-      if major_minor_p ~= major_minor_np then
-        return false, "plugin \"" .. p.name .. "\" version incompatible, " ..
-                      "CP version: " .. tostring(p.version) ..
-                      " DP version: " .. tostring(np.version) ..
-                      " DP plugin version acceptable is "..
-                      major_minor_p .. ".x",
-                      CLUSTERING_SYNC_STATUS.PLUGIN_VERSION_INCOMPATIBLE
+
+function _M:check_configuration_compatibility(dp_plugin_map)
+  for _, plugin in ipairs(self.plugins_list) do
+    if self.plugins_configured[plugin.name] then
+      local name = plugin.name
+      local cp_plugin = self.plugins_map[name]
+      local dp_plugin = dp_plugin_map[name]
+
+      if not dp_plugin then
+        if cp_plugin.version then
+          return nil, "configured " .. name .. " plugin " .. cp_plugin.version ..
```

Comment:
ditto, let's not make two different logs here. the version of CP plugins doesn't actually
make a difference to the compat check here

Reply:
Single cp always logs the same, the only difference in these lines is that other contains version while other does not.